design/vela-core/definition-versioning.md

```diff
@@ -0,0 +1,321 @@
+# Versioning Support for KubeVela Definitions
+
+<!-- toc -->
+- [Versioning Support for KubeVela Definitions](#versioning-support-for-kubevela-definitions)
+  - [Summary](#summary)
+  - [Scope](#scope)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Acceptance Criteria](#acceptance-criteria)
+  - [Current Implementation](#current-implementation)
+    - [Versioning](#versioning)
+    - [Auto Upgrade](#auto-upgrade)
+      - [Reference:](#reference)
+  - [Proposal](#proposal)
+    - [Details](#details)
+    - [Issues](#issues)
+  - [Examples](#examples)
+<!-- /toc -->
+
+## Summary
+
+Support Semantic versioning for KubeVela Components and a way to allow fine control over auto-upgrades of KubeVela Applications to new versions of a Component. The implementation should include support for consistent versioning across environments/clusters, meaning specific Revisions/Versions of a Component should have consistent behaviour.
+
+## Scope
+
+Although, this document limits the scope of discussion to ComponentDefinition Revisions/Versions, due to the current implementation, the changes will most likely apply to all [`Definition`](https://kubevela.io/docs/getting-started/definition/) types. These changes are planned to be explored and validated as part of the implementation.
+
+## Motivation
+
+OAM/KubeVela Definitions (referred to as ComponentDefinitions of Components in the rest
+of the document) are the basic building blocks of the KubeVela platform. They
+expose a contract similar to an API contract, which evolves from minor to major
+versions. Applications are composed of Components that the KubeVela engine stitches
+together.
+
+KubeVela creates a `DefinitionRevision` for all changes in a Component `spec`.  
+Currently, Applications can refer to a particular Revision of a Component. 
+But, this versioning scheme has the following issues: 
+
+- The `DefinitionRevision` does not denote the type of the change (patch/bug, minor or major). This hinders automation of automatic upgrades.
+- The current scheme also doesn't allow much control over automatic upgrades to new Component Revisions. KubeVela automatically upgrades/reconciles the Application to the
+latest when no Component Revision is specified. 
+  > While we don't ideally want Application developers to bother with such details, there are use cases where
+  > an automatic upgrade to the latest Component version is not desired.
+
+
+### Goals
```

Comment:
Does this design allow a definition author to support 2 separate major versions of a definition simultaneously? In other words, are the following use cases supported?

>**AS A** Component author\
>**I SHOULD** be able to publish a new major version of my Component (v1.x -> v2.x)\
>**SO THAT** I can introduce breaking changes to my definition (in v2.x) without breaking existing consumers (using v1.x).

>**AS A** Component author\
>**I SHOULD** be able to publish changes for 2 different major versions of my Component simultaneously (v1.x, v2.x)\
>**SO THAT** I can continue updating the old version (v1.x) until consumers have migrated to the new one (v2.x).

If yes, how?

Let's say a definition has revision v1.0.0 and v2.0.0, and the author wants to publish v1.0.1 and v2.0.1.

How would this be done? Would the author sequentially apply these changes to the same ComponentDefinition, once for v1.0.1 and once for v2.0.1? Would the author directly apply a DefinitionRevision each for v1.0.1 and v2.0.1?

Reply:
Both scenarios are supported. Please refer the [documentation PR](https://github.com/kubevela/kubevela.github.io/pull/1343) for more information.